aten/src/ATen/core/ivalue.cpp

```diff
@@ -604,4 +604,45 @@ getClassConverter() {
       classConverter;
   return classConverter;
 }
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    List<intrusive_ptr<ivalue::Future>> srcs) {
+  struct Ctx {
+    explicit Ctx(List<intrusive_ptr<ivalue::Future>> srcs)
+        : remaining(srcs.size()),
+          srcFutures(std::move(srcs)),
+          asIvalue(srcFutures),
+          dstFuture(make_intrusive<ivalue::Future>(asIvalue.type())) {}
+    std::atomic<int32_t> remaining{0};
+    List<intrusive_ptr<ivalue::Future>> srcFutures;
+    IValue asIvalue;
+    intrusive_ptr<ivalue::Future> dstFuture;
+  };
+
+  auto ctx = std::make_shared<Ctx>(std::move(srcs));
+  std::function<void()> func = [ctx]() {
+    if (--ctx->remaining == 0) {
+      ctx->dstFuture->markCompleted(ctx->asIvalue);
+    }
+  };
+  for (int32_t tot = ctx->srcFutures.size(), i = 0; i < tot; ++i) {
+    ctx->srcFutures.get(i)->addCallback(func);
+  }
+  if (ctx->srcFutures.size() == 0) {
+    ctx->dstFuture->markCompleted(ctx->asIvalue);
+  }
+  return ctx->dstFuture;
+}
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    std::vector<intrusive_ptr<ivalue::Future>> srcs) {
+  auto typePtr =
+      !srcs.empty() ? srcs[0]->type() : FutureType::create(NoneType::get());
```

Comment:
This means all `Future`s in `srcs` need to be the same type? Do we need to `TORCH_CHECK` that?

Reply:
Yes, this is effectively (and surprisingly) an artifact of the current `List<>` and `Future<>` apis.
I'll add a `TORCH_CHECK()`.

`c10::List` requires every entry to have the same underlying type, and `Future<Any_IValue>` isn't quite possible - a specific `TypePtr` is required. I was a bit surprised by some of the constraints imposed by the IValue library.

Reply:
(not requesting changes to this PR) I wonder if it is possible to support collecting a list of Futures with different types?

Reply:
actually, I didn't realize that there was `AnyType` in `jit_type.h`
If I change typeptr to `FutureType::create(AnyType::get());` things seem to work, and I don't need any of these checks, which is much better.

Reply:
update (also in comment in other area) is that I realized that ` FutureType::create(AnyType::get())` seems to work, which is better